src/LanguageServer/Impl/Sources/DocumentHighlightSource.cs

```diff
@@ -0,0 +1,65 @@
+﻿// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABILITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Python.Analysis;
+using Microsoft.Python.Analysis.Documents;
+using Microsoft.Python.Analysis.Modules;
+using Microsoft.Python.Analysis.Types;
+using Microsoft.Python.Core;
+using Microsoft.Python.Core.IO;
+using Microsoft.Python.Core.Text;
+using Microsoft.Python.LanguageServer.Documents;
+using Microsoft.Python.LanguageServer.Protocol;
+
+namespace Microsoft.Python.LanguageServer.Sources {
+    internal sealed class DocumentHighlightSource {
+        private const int DocumentHighlightAnalysisTimeout = 10000;
+        private readonly IServiceContainer _services;
+
+        public DocumentHighlightSource(IServiceContainer services) {
+            _services = services;
+        }
+
+        public async Task<DocumentHighlight[]> DocumentHighlightAsync(Uri uri, SourceLocation location, CancellationToken cancellationToken = default) {
+            if (uri != null) {
+                var analysis = await Document.GetAnalysisAsync(uri, _services, DocumentHighlightAnalysisTimeout, cancellationToken);
```

Comment:
One thing we might want to think about is returning `null` if there is no analysis yet, to allow the editor to fall back to lazy. Right now, this call will end up returning an empty list and VS Code doesn't highlight anything.

Reply:
This might be a blocking thing, actually, since the experience regresses until the analysis is complete. I _think_ it would be as simple as `if (analysis is EmptyAnalysis) { return null }` but @MikhailArkhipov would know best how to check that things aren't ready.

Reply:
This is beyond my knowledge for the moment. However when I'm testing this in Emacs, I get highlighting while the server is analyzing.

Reply:
It would also be nice if the analysis could provide information whether the symbol is read or write, but this is just a "nice to have" feature.

Reply:
Previous analysis may or may not be empty, if timeout expires waiting on analysis, previous one will be returned. One possible way to determine if analysis is stale is to check `Document.Version` against `Analysis.Version`

Reply:
As for symbol R/W do you mean it can be renamed/changed? There is a root definition of the symbol that `Rename` operation uses which has URI of the defining module. Via `IRunningDocumentsTable` aka `RDT` one can fetch the document and check if `Document.ModuleType == ModuleType.User` since symbols in other module types are read-only.

Reply:
R/W meaning "this location reads from `x`" versus "this location writes to `x`".

Reply:
`Document.Version` and `Analysis.Version` are incremented separately, so comparing them wouldn't work. We would need to convert entirely to the snapshot model with a single incrementing version across everything for that to work (which would be awesome, along with using LSP's versioned identifiers).

I just re-tested, and maybe it doesn't break and we can just leave the PR as-is. I'm still testing to make sure.

Reply:
👍

Reply:
I meant as Jake interpreted it ;-). DocumentHighlight can tag a symbol instance as "Read", "Write" or "Text". In the following simple example "x" is redefined in the last line which means that it could be tagged as "Read". Emacs LSP renders "Read" symbols in italics. But this is not a showstopper for me.

![lsp](https://user-images.githubusercontent.com/43724994/68799664-d3e35080-0658-11ea-9908-cb3ec4b0da89.png)

Reply:
@astenman Perhaps I'm confused, but doesn't your PR include that information? Or are you using Read/Write a different way?